#!/bin/bash

################################################################################
# Author: T. Cameron Waller
# Date, first execution: 15 March 2023
# Date, last execution: __ March 2023
# Review: TCW; ___
################################################################################
# Note

# This script calls a method within PLINK2 to calculate a polygenic score as
# linear combination of allelic effects across SNPs within a genotype.
# https://www.cog-genomics.org/plink/1.9/score
# https://www.cog-genomics.org/plink/2.0/score

# SNP effects
# Description: Format of SNP effects for calculation of polygenic scores
# Description: This format attempts to be compatible with team standard for GWAS summary statistics.
# Documentation site: https://www.cog-genomics.org/plink/1.9/score
# Documentation site: https://www.cog-genomics.org/plink/2.0/score
# File suffix: ".txt.gz"
# File type: text
# File compression: GZip
# Delimiter: white space
# Header: Yes
# Chromosome base position coordinate system: base 1
#   Site: https://www.biostars.org/p/84686/
#   Note: Coordinates designate 1-based integer position of each base
# Columns: SNP CHR BP A1 A2 A1AF A1EFFECT SE P
#          1   2   3  4  5  6    7        8  9

# Polygenic scores
# Description: output file of PLINK2 "score" command is in ".sscore" format
# Documentation site: https://www.cog-genomics.org/plink/2.0/formats#sscore
# File suffix: ".sscore"

################################################################################
################################################################################
################################################################################

# TODO: TCW; 17 August 2022
# TODO: as the PRS-CSX documentation says, "If polygenic scores are generated by
# TODO: chromosome, use the 'sum' modifier so that they can be combined into a genome-wide score."

# TODO: TCW; 18 August 2022
# TODO: This script will accept a SINGLE file of posterior effect size report from PRS-CSX.
# TODO: This script will call PLINK2 --score for this SINGLE file.
# TODO: This script will need a driver script. Follow pattern of "3_estimate_prscs_allelic_effects.sh"


################################################################################
################################################################################
################################################################################


################################################################################
# Organize arguments.
path_file_source_effects=${1} # full path to source file in standard format of allelic effects across SNPs
path_file_source_genotypes=${2} # full path to source file in Variant Call Format (VCF) of target genotypes
path_file_product=${3} # full path to product file for polygenic scores of allelic effects across target genotypes
threads=${4} # count of processing threads to use
report=${5} # whether to print reports

###########################################################################
# Organize paths.

# Directories.
cd ~/paths
path_plink2=$(<"./tools_plink2.txt")

path_directory_product="$(dirname $path_file_product)"
name_base_file_product="$(basename $path_file_product .txt.gz)"
path_directory_product_temporary="${path_directory_product}/temporary_${name_base_file_product}" # hopefully unique

# Files.
path_file_temporary_sscore="${path_directory_product_temporary}/${name_base_file_product}.sscore"

# Initialize directory.
#rm -r $path_directory_product
rm -r $path_directory_product_temporary
mkdir -p $path_directory_product
mkdir -p $path_directory_product_temporary
cd $path_directory_product_temporary


################################################################################
# Activate Virtual Environment.
# Read private, local file paths.
#echo "read private file path variables and organize paths..."
cd ~/paths
path_plink2=$(<"./tools_plink2.txt")

################################################################################
# Call PLINK2 to calculate linear combination of allelic effects across SNPs in
# target genotypes.

# PLINK2 arguments "center", "variance-standardize", and "dominant" modify the
# scale of the polygenic scores; however, wait to adjust scale until after
# combination of scores across separate chromosomes.

$path_plink2 \
--memory 90000 \
--threads $threads \
--vcf $path_file_source_genotypes \
--xchr-model 2 \
--score $path_file_source_effects 1 4 header no-mean-imputation ignore-dup-ids list-variants \
--score-col-nums 7 \
--out $name_base_file_product

# Compress file format.
gzip -cvf $path_file_temporary_sscore > $path_file_product

# Remove temporary, intermediate files.
#rm -r $path_directory_product_temporary



#
